Add tanh_cuda support for complex types

krshrimali opened the pull request:
Similar to #38400, this PR atempts to add support for complex types for `tanh` (CUDA).

cc: @anjali411

krshrimali committed:
torch.prod works fine now with fp16 input and fp32 output tensor

krshrimali committed:
torch.prod works fine now with fp16 input and fp32 output tensor

krshrimali committed:
Replicates sum_kernel_cuda and sum_kernel_impl, adds out_t arg

krshrimali committed:
Replicates sum_kernel_cuda and sum_kernel_impl, adds out_t arg

krshrimali committed:
Test added for test_prod_gpu, under review, passes with changes successfully

krshrimali committed:
Test added for test_prod_gpu, under review, passes with changes successfully

krshrimali committed:
modified test for torch.prod (only CUDA), uses all dtype combinations

krshrimali committed:
modified test for torch.prod (only CUDA), uses all dtype combinations

krshrimali committed:
use torch.tensor.dtype instead of torch.tensor.type()

krshrimali committed:
use torch.tensor.dtype instead of torch.tensor.type()

krshrimali committed:
Merge branch 'master' of https://github.com/krshrimali/pytorch

krshrimali committed:
:Merge branch 'master' of https://github.com/pytorch/pytorch

krshrimali committed:
Merge remote-tracking branch 'upstream/master'

krshrimali committed:
adding test_torch back to master

anjali411 commented on aten/src/ATen/native/cuda/UnaryGeometricKernels.cu:
```diff
@@ -150,15 +152,22 @@ void cosh_kernel_cuda(TensorIterator& iter) {
+}
+
 void tanh_kernel_cuda(TensorIterator& iter) {
   AT_DISPATCH_FLOATING_TYPES_AND2(ScalarType::Half, ScalarType::BFloat16, iter.dtype(), "tanh_cuda", [&]() {
```
`AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2`

krshrimali replied:
Sorry, my bad. Missed this. Should be resolved here: https://github.com/pytorch/pytorch/pull/38486/commits/a930fce4ae0637e5a178a0c124efa9ffd98670f7. Thanks!

anjali411 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -246,6 +246,7 @@ def method_tests():
         ('log2', uniform_scalar(1e-2j, requires_grad=True), NO_ARGS, 'complex_scalar', (True,)),
         ('tanh', (S, S, S), NO_ARGS, '', (True,)),
         ('tanh', (), NO_ARGS, 'scalar', (True,)),
+        ('tanh', (S, S, S), NO_ARGS, 'complex', (True,)),
```
you should include `tanh` in this [list](https://github.com/pytorch/pytorch/blob/7f11079769b0c9cb678a1046f1b0c9259ca3a3ac/test/test_autograd.py#L4193)

krshrimali replied:
Do you mean [here](https://github.com/pytorch/pytorch/blob/c06bd9289c9b73b200a4275a6ea6c34e8129ed4a/test/test_autograd.py#L4193)? I think I already included `tanh` in the list. Please correct me if I'm getting anything wrong here. Thanks!

anjali411 replied:
yeah sorry my b for some reason that change didn't show up for me last time

anjali411 commented:
I see test failure for test_autograd `RuntimeError: "tanh_backward_cpu" not implemented for 'ComplexDouble'` so you would need to update `tanh_backward_kernel` as well for the autograd test to run fine for complex

anjali411 reviewed (changes requested):
let's update the `tanh_backward_kernel` and `tanh_backward_kernel_cuda`

krshrimali commented:
Thanks @anjali411! Didn't realize that it was "backward" (missed the keyword there). I've pushed the changes, can you please review the PR again?

anjali411 commented:
LGTM there's a merge conflict so just rebase it and should be fine

krshrimali commented:
@anjali411 - can you please review the changes? Please let me know if there is anything left for this PR. :)

anjali411 commented on aten/src/ATen/native/cuda/UnaryGeometricKernels.cu:
```diff
@@ -150,15 +152,22 @@ void cosh_kernel_cuda(TensorIterator& iter) {
-  AT_DISPATCH_FLOATING_TYPES_AND2(ScalarType::Half, ScalarType::BFloat16, iter.dtype(), "tanh_cuda", [&]() {
-    AT_SKIP_BFLOAT16_IF_NOT_ROCM(scalar_t, "tanh_cuda", [&] {
-      gpu_kernel(iter, []GPU_LAMBDA(scalar_t a) -> scalar_t {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(ScalarType::Half, ScalarType::BFloat16, iter.dtype(), "tanh_cuda", [&]() {
```
`AT_DISPATCH_FLOATING_AND_C10_COMPLEX_TYPES_AND2`

krshrimali replied:
I don't see this macro defined in the master branch. Have added this macro though.

anjali411 commented on aten/src/ATen/native/cuda/UnaryGeometricKernels.cu:
```diff
@@ -150,15 +152,22 @@ void cosh_kernel_cuda(TensorIterator& iter) {
-    AT_SKIP_BFLOAT16_IF_NOT_ROCM(scalar_t, "tanh_cuda", [&] {
-      gpu_kernel(iter, []GPU_LAMBDA(scalar_t a) -> scalar_t {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(ScalarType::Half, ScalarType::BFloat16, iter.dtype(), "tanh_cuda", [&]() {
+    using thrust_t = typename ztype_cuda<scalar_t>::thrust_t;
```
we can remove this once the above change is made.

krshrimali replied:
Done!

krshrimali replied:
As per offline (slack) discussion, this has been reverted back from using `c10::complex`. Eventually though, we'll use c10::complex once we have the support.

anjali411 commented on aten/src/ATen/native/cuda/UnaryGeometricKernels.cu:
```diff
@@ -150,15 +152,22 @@ void cosh_kernel_cuda(TensorIterator& iter) {
 }
 
+template<typename T>
+__host__ __device__ static inline thrust::complex<T> tanh_wrapper(thrust::complex<T> v) {
```
`__host__ __device__ static inline c10::complex<T> tanh_wrapper(c10::complex<T> v)`

krshrimali replied:
Done!

krshrimali replied:
Reverted back to `thrust::complex<T>` till the time full migration from `std::complex` to `c10::complex` is complete. :)

anjali411 commented:
left some minor comments.
test failures look flaky but let's rebase and see the CI

krshrimali commented:
Screwed things up while rebasing. Attempting to fix this.

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into tanh-complex-support-cuda

krshrimali committed:
Adding tanh support for cuda, tests added

krshrimali committed:
tanh backward support (cuda) for complex types

krshrimali commented:
Things should be back to normal now. 

* Changes to `tanh_backward_kernel` (cpu) have been ommitted since a PR adding these changes (#37791) was merged recently.
* Changes to `tanh_backward_kernel_cuda` are done, but using macro: `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES`. Please let me know if this needs to be changed.
* We are now using `c10::complex` and appropriate macros for `c10::complex` in `aten/src/ATen/native/cuda/UnaryGeometricKernels.cu`. But, I don't see the macro defined in the master branch (`aten/src/ATen/Dispatch.h`). I'm not aware if there is a PR to add this macro.
* Updated tests. #37791 used `expectedFailureCUDA` in `torch/testing/_internal/common_methods_invocations.py` since there was no support for complex types (CUDA) for `tanh` by then. Removed the keyword from there.

cc: @anjali411

krshrimali committed:
Revert back to thrust::complex

krshrimali committed:
Change macro

anjali411 commented on aten/src/ATen/native/cuda/BinaryMiscBackwardOpsKernels.cu:
```diff
@@ -20,7 +20,7 @@ void sigmoid_backward_kernel_cuda(TensorIterator& iter) {
 
 void tanh_backward_kernel_cuda(TensorIterator& iter) {
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.dtype(), "tanh_backward_cuda", [&]() {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(iter.dtype(), "tanh_backward_cuda", [&]() {
```
`AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(ScalarType::Half, iter.dtype() ...)`

krshrimali replied:
Should be resolved here: https://github.com/pytorch/pytorch/pull/38486/commits/a38b4b34a2a70d9498ae25cddb3199805072449a. Thanks!

anjali411 commented:
let's add the dispatch for tanh for Half back again and it should be good

krshrimali committed:
Add half support for tanh backward (cpu) back

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into tanh-complex-support-cuda

anjali411 commented on aten/src/ATen/native/cuda/BinaryMiscBackwardOpsKernels.cu:
```diff
@@ -20,7 +20,7 @@ void sigmoid_backward_kernel_cuda(TensorIterator& iter) {
 void tanh_backward_kernel_cuda(TensorIterator& iter) {
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.dtype(), "tanh_backward_cuda", [&]() {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(ScalarType::Half, iter.dtype(), "tanh_backward_cuda", [&]() {
     gpu_kernel(iter, []GPU_LAMBDA(scalar_t a, scalar_t b) -> scalar_t {
```
`using thrust_t = typename ztype_cuda<scalar_t>::thrust_t;`

krshrimali replied:
Should be resolved here: https://github.com/pytorch/pytorch/pull/38486/commits/c36a4480497346ef3a831673a50be5d316645e33. Thanks!

krshrimali committed:
use thrust_t

krshrimali committed:
Merge branch 'master' into tanh-complex-support-cuda

krshrimali commented:
Please take a look, @anjali411. There were a few merge conflicts, which I have resolved. The 2 failures reported by CI seem unrelated to the changes in this PR.

krshrimali commented:
Closing this PR as another PR added complex support. Thanks!
